python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -621,6 +592,68 @@ def _get_lookup_hash(self, config_uri, project, entity_type):
             entity_type=entity_type
         )
 
+    def _get_pipeline_configuration_data(self, manager, project_entity, data):
+        """
+        Gathers all of the necessary data pertaining to the project's pipeline
+        configurations. This includes the PipelineConfiguration entity, the
+        contents and lookup hashes, and the associated descriptor object.
+
+        :param manager: A ToolkitManager.
+        :param dict project_entity: The Project entity dict.
+        :param dict data: The payload from the client.
+
+        :returns: A dictionary, keyed by PipelineConfiguration entity id, that
+            contains dictionaries with "contents_hash", "lookup_hash",
+            "descriptor", and "entity" keys.
+        :rtype: dict
+        """
+        if "config_data" in self.WSS_KEY_CACHE[self._wss_key]:
+            self.logger.debug("Pipeline config data found for %s", self._wss_key)
+        else:
+            config_data = dict()
+            pcs = self._get_pipeline_configurations(
+                manager,
+                project_entity,
+            ) or [dict(id=None, name="Primary")]
+
+            for pc in pcs:
+                self.logger.debug("Processing config: %s" % pc)
+
+                # The hash that acts as the key we'll use to look up our cached
+                # data will be based on the entity type and the pipeline config's
+                # descriptor uri. We can get the descriptor from the toolkit
+                # manager and pass that through along with the entity type from SG
+                # to the core hook that computes the hash.
+                manager.pipeline_configuration = pc["id"]
+
+                try:
+                    pc_descriptor = manager.resolve_descriptor(project_entity)
+                except sgtk.bootstrap.TankBootstrapError as exc:
+                    self.logger.debug("Unable to resolve config descriptor, skipping: %s" % pc)
+                    self.logger.debug(str(exc))
+                    continue
+
+                self.logger.debug("Resolved config descriptor: %r" % pc_descriptor)
+                pc_key = pc_descriptor.get_uri()
+
+                pc_data = dict()
+                pc_data["contents_hash"] = self._get_contents_hash(
+                    pc_descriptor,
+                    self._get_site_state_data(),
+                )
+                pc_data["lookup_hash"] = self._get_lookup_hash(
+                    pc_key,
+                    project_entity,
+                    data["entity_type"],
+                )
+                pc_data["descriptor"] = pc_descriptor
+                pc_data["entity"] = pc
+                config_data[pc["id"]] = pc_data
+
+            self.WSS_KEY_CACHE[self._wss_key]["config_data"] = config_data
```

Comment:
some more information about when this cache validates would be nice! both in the code, and i am also curious to understand more!